localstack/utils/archives.py

```diff
@@ -185,12 +185,29 @@ def download_and_extract(archive_url, target_dir, retries=0, sleep=3, tmp_archiv
     if not os.path.exists(tmp_archive) or os.path.getsize(tmp_archive) <= 0:
         # create temporary placeholder file, to avoid duplicate parallel downloads
         save_file(tmp_archive, "")
-        for i in range(retries + 1):
+
+        current_try = 1
+        while current_try <= retries + 1:
             try:
                 download(archive_url, tmp_archive)
                 break
-            except Exception:
-                time.sleep(sleep)
+            except Exception as e:
+                LOG.warning(
+                    "Attempt %d. Failed to download archive from %s: %s",
+                    current_try,
+                    archive_url,
+                    e,
+                )
+                # only sleep between retries, not after the last one
+                if current_try <= retries:
+                    time.sleep(sleep)
+            finally:
+                current_try += 1
+
+    # if the temporary file we created above hasn't been replaced, we assume failure
+    if os.path.getsize(tmp_archive) <= 0:
+        raise Exception("Failed to download archive from %s: . Retries exhausted", archive_url)
```

Comment:
TBH, just copied this from L185.

Like you said, probably not, but I guess we still wouldn't want it if for *some* reason this would happen on some esoteric system :man_shrugging: